test/state/player_state_test.cpp

```diff
@@ -161,3 +162,144 @@ TEST_F(PlayerStateTest, GetOffsetFromPosition) {
+    bot2.id = 29;
+    bot2.position = {2, 2};
+    player_state::Bot bot3(bot2);
+
+    EXPECT_EQ(bot1.id, 11);
+    EXPECT_EQ(bot1.final_destination, DoubleVec2D::null);
+    EXPECT_EQ(bot1.transform_destination, DoubleVec2D::null);
+    EXPECT_EQ(bot1.transforming, false);
+    EXPECT_EQ(bot1.blasting, false);
+    EXPECT_EQ(bot1.destination, DoubleVec2D::null);
+    EXPECT_EQ(bot1.position, DoubleVec2D(0, 0));
+
+    // == & != operator checks
+    EXPECT_EQ(bot2, bot3);
+    EXPECT_NE(bot1, bot2);
+
+    // target_position!=position case
+    bot1.blast({10, 10});
+    EXPECT_EQ(bot1.final_destination, DoubleVec2D(10, 10));
+    EXPECT_EQ(bot1.blasting, false);
+
+    bot1.reset();
+    EXPECT_EQ(bot1.transform_destination, DoubleVec2D::null);
+    EXPECT_EQ(bot1.final_destination, DoubleVec2D::null);
+    EXPECT_EQ(bot1.transforming, false);
+    EXPECT_EQ(bot1.destination, DoubleVec2D::null);
+    EXPECT_EQ(bot1.blasting, false);
+
+    // transforms test on bot1
+    bot1.position = {9, 9};
+    bot1.transform({9, 9});
+    EXPECT_EQ(bot1.transforming, true);
+
+    // target_position==postion case on bot2
+    EXPECT_EQ(bot2.position, DoubleVec2D(2, 2));
+    bot2.blast({2, 2});
+    EXPECT_EQ(bot2.transforming, false);
+    EXPECT_EQ(bot2.transform_destination, DoubleVec2D::null);
+    EXPECT_EQ(bot2.final_destination, DoubleVec2D::null);
+    EXPECT_EQ(bot2.position, DoubleVec2D(2, 2));
+    EXPECT_EQ(bot2.blasting, true);
+
+    bot2.reset();
+    EXPECT_EQ(bot2.transform_destination, DoubleVec2D::null);
+    EXPECT_EQ(bot2.final_destination, DoubleVec2D::null);
+    EXPECT_EQ(bot2.transforming, false);
+    EXPECT_EQ(bot2.destination, DoubleVec2D::null);
+    EXPECT_EQ(bot2.blasting, false);
+
+    bot2.transform({10, 10});
+    EXPECT_EQ(bot2.transform_destination, DoubleVec2D(10, 10));
+    EXPECT_EQ(bot2.transforming, false);
+
+    bot2.transform();
+    EXPECT_EQ(bot2.transforming, true);
+}
+
+TEST_F(PlayerStateTest, MapElementTest) {
+    auto &elem1 = player_map[1][2];
+    elem1.setTerrain(player_state::TerrainType::LAND);
+    EXPECT_EQ(elem1.getTerrain(), player_state::TerrainType::LAND);
+}
+
+TEST_F(PlayerStateTest, OverloadOperatorsTest) {
+    std::ostringstream result;
+    std::ostringstream expected;
+
+    // Tower and TowerState ostream overloads test
+    auto &tower = player_states[0].towers[1];
+    tower.id = 10;
+    tower.state = TowerState::DEAD;
+    // towerstate ostream << test
+    result << tower.state;
+    EXPECT_EQ(result.str(), string("DEAD"));
+    result.str(""); // clears the stream
+    result.clear();
+    // << for Tower class
+    expected << "Tower(id: " << tower.id << ") {" << endl;
+    expected << "   hp: " << tower.hp << endl;
+    expected << "   state: " << tower.state << endl;
+    expected << "}" << endl;
+    result << tower;
+    EXPECT_EQ(result.str(), expected.str());
+    result.str("");
+    result.clear();
+    expected.str("");
+    expected.clear();
```

Comment:
Why are you making result and expected as empty strings before clearing them?

Reply:
.str("") will clear the buffer to have just the empty string and clear() will remove all the error flags if they had been set during the writes before, so the empty string part is needed.

Reply:
Okay, just add a short comment mentioning what clear is doing